### DISCUSSION TITLE
Prefer importing using module/package/component prefix

### DISCUSSION THREAD

**Discussion by oliv3r** [home-assistant/architecture#953] [General]
Dear all,

this is an architectural proposal to (hopefully) help improve readability and maintainability of HASS code with regards to imports and the imported objects thereof.

Lets start with an example first.

As I was trying to learn about the recorder module, I found that not only are we applying import methodologies inconsistently, the way we do it is also for some bits hard to find the uses. Also, the recorder is 'interesting' as quite a few components import it to do recording of data.

The recorder seems to imported for the most part as:

```python
from homeassistant.components.recorder import (
    DOMAIN,
    get_instance,
)
```

This means, that (at least within this module) we have to know that DOMAIN, is in fact, recorder.DOMAIN, as is get_instance, really recorder.get_instance.

But we also see the following form:

```python
from homeassistant.components.recorder import (
    DOMAIN as RECORDER_DOMAIN,
    get_instance as get_recorder_instance,
)
```

While the above avoids the namespacing clashing, there is no rules on naming. This becomes very messy very quickly and makes grepping for all the correct instances extremly hard.

But we do not follow tihs paradigm all the time. After searching through the various imports, quite a few imports are being handled via the 'generally more appropriate' way (for recorder this isn't done often, but it IS being done', which results in the following import and usage.

```python
from homeassistant.components import recorder

...
instance = recorder.get_instance()
if domain = recorder.DOMAIN
```

The obvious advantages I personally can see:

* readability: It makes it immediately clear, what component get_instance belongs to, and what we are getting the instance for, without using non-standard or without rules, for renaming (using AS) things.
* context: Like the above, when reading code, the context of a piece of code also becomes much more obvious and clear, self-documenting code if you will.
* searchability: If one wants to know all users of `get_instance` of the recorder, grep (or your IDE of choice's search function) can much more easily and much more accurately find all correct references. Personally, I feel quite strongly about the grep-ability.
* consistency: Using dot-notation is consistent with Python's naming conventions (as outlined in PEP 8) and the overall design philosophy of the language.
* Avoid naming conflicts: Considering the size of home-assistant (and the number of users of the recording module as part of this example), makes it much harder to have namespace conflicts
* IDE integration: Using dot-notation allows IDE's to help with autocompletion
* refactoring: ties in with searchability, but having dot-notation allows for near automatic refactoring

The major downside I can think of, is it requires more typing. Which imo gets slightly compensated by auto-complete; and personally I'm much against this paradime. You write code once, but read it multiple times. Being more explicit in this sense (within reason) imo is always better.

Looking at PEP020 to help with some thoughts

* Beautiful is better than ugly
   - This is obvious up for debate. Personally, I find the suggestions made above more beautiful :)
* Explicit is better than implicit
   - I think it is obvious that this is what we are trying to here ...
* Simple is better than complex
   - While its slightly more typing, I think it does keep things more simple, at least mentally
* Sparse is better than dense
   - Basically, the more typing vs less typing
* Readability counts
   - I think this improves readability
* Special cases aren't special enough to break the rules
   - the 'import as' makes it seem like its special, but its not really
* Although practicality beats purity
   - I do NOT think this makes it more practical, just to type a bit less
* In the face of ambiguity, refuse the temptation to guess
  - not using the dot-notation prefix makes it actual more ambiguous
* There should be one-- and preferably only one --obvious way to do it
  - By avoiding treating it special, it is quite an obvious way to do it
* Although that way may not be obvious at first unless you're Dutch
  - I am dutch, so this is quite obvious :D
* Now is better than never
  - https://github.com/home-assistant/core/pull/96835 :p
* If the implementation is hard to explain, it's a bad idea
   - I have tried my best to explain the reasons, and it wasn't hard at all
   - Having to explain/warrant why it's a good idea to not use a dot-notation is imo much harder, as I can't think of a good reason even.
* If the implementation is easy to explain, it may be a good idea
   - See above :D
* Namespaces are one honking great idea -- let's do more of those!
  - This is probably key, yeah? Which is what this entire post is about I'd argue!!

When asking both chatgpt and bard (yeah I know, not super trustworthy resources :D) about this, they both agree, that within their available knowledge. A few snippets from their replies:
```
The experience of Python developers. Many experienced Python developers recommend that you reference Python objects or modules by their prefix. They argue that this makes code more readable, understandable, and maintainable.

I have also seen this best practice used in many Python codebases. This suggests that it is a widely accepted practice.

Overall, using dot-notation to access namespaces aligns with the Pythonic approach of writing clean, readable, and maintainable code. It takes advantage of the language's module system, helps prevent naming conflicts, and fosters good coding practices within the Python community.
```

But lets not take these hallucinating tools for their word, but try to argue against or for it. Lets see what the community things :) Or things against it.

**Comment by frenck** [comment]
I'm not in favor of importing complete modules like that. The advantages you list, I personal don't agree with and raise the same ones for the other method:

> The obvious advantages I personally can see:

> readability: It makes it immediately clear, what component get_instance belongs to, and what we are getting the instance for, without using non-standard or without rules, for renaming (using AS) things.

I don't agree, it adds more smush throughout the code, making it harder to read. I personally become blinded by all the extra stuff not needed.

> context: Like the above, when reading code, the context of a piece of code also becomes much more obvious and clear, self-documenting code if you will.

Not true, in some cases maybe. But the context can be lost as well, as we have lots of modules that have the same name. For example, a recorder can be an entity platform implementation too. The context doesn't tell you the origin of `recorder` in the given example.

> searchability: If one wants to know all users of get_instance of the recorder, grep (or your IDE of choice's search function) can much more easily and much more accurately find all correct references. Personally, I feel quite strongly about the grep-ability.

Same as above. It doesn't solve it. There are many tools that can help you a lot better with that instead of searching. That said, you can still search for the import in the current situation.
IDEs will pick up the references as well.

> consistency: Using dot-notation is consistent with Python's naming conventions (as outlined in PEP 8) and the overall design philosophy of the language.

Not really, so does the current 🤷 

> Avoid naming conflicts: Considering the size of home-assistant (and the number of users of the recording module as part of this example), makes it much harder to have namespace conflicts

That will still exists, as pointed out above.

> IDE integration: Using dot-notation allows IDE's to help with autocompletion

This is a non-reason. IDEs have no issues with the existing method either, as a matter of fact, they will even update the imports for you nowadays (and even clean them up when you are no longer using an imported reference).

> refactoring: ties in with searchability, but having dot-notation allows for near automatic refactoring

That is really a duplicate. Refactoring in IDEs don't care and will work either case.

> Looking at PEP020 to help with some thoughts

PEP020 is the biggest **&**% Python ever produced, I don't see how they ever where able to write that as the text it conflicting itself, nor do they take it into account for many language decisions they have made in the past. That said, I see not technical relation to anything in this proposal to that PEP.


TL;DR: I'm not in favor. If anything, I would raise the opposite.

../Frenck

**Comment by thecode** [comment]
I think this is a clean way and prefer it where possible:
```python
from homeassistant.components.recorder import (
    DOMAIN,
    get_instance,
)
```

I support making guidelines (not rules) for recommended naming in this case:
```python
from homeassistant.components.recorder import (
    DOMAIN as RECORDER_DOMAIN,
    get_instance as get_recorder_instance,
)
```

I don't like importing complete package, I prefer to know what is used from the package when looking at the imports.

IDE issues were already answered by Frenck, I never experienced any problems with any of the options above, the IDE (vscode in my case) handles renames, imports, autocompletion correctly in all cases.

**Reply by oliv3r** [reply]
> I don't like importing complete package, I prefer to know what is used from the package when looking at the imports.

So this is far deeper then what my little python knowledge holds, so you may know this much better, but from what I understood;

yes 'from X import *' actually imports everything, and is considered bad. But afaik from 'X import object' doesn't import 'everything', just the object or rather even tells the interpreter how to resolve the invocation of a method of the object.

> I support making guidelines (not rules) for recommended naming in this case

from a naming pov, this is an interesting case you make. you PREFIX the domain with the name of the module (good) but then prefer to have the module name 'somewhere in the middle' :) so getting this consistently right with a simple rule is hard, I would in that case favor `recorder_get_instance` to keep things clear and consistent.

**Comment by oliv3r** [comment]
I see this is getting quite heavily loaded and very opinionated :)

Obviously style tends to be quite personal, and when vested in a single code base, one gets accustomed to that style and is very much not in favor of changing from that.

Also, on the arguments, it's very easy to go back and forth likewise.

Anyway, I'll try to respond anyway, just so that it was said.

> I don't agree, it adds more smush throughout the code, making it harder to read. I personally become blinded by all the extra stuff not needed.

This is obviously very personal. You are very deeply invested in the code base, but everybody else, that doesn't spend days in the code base. So then the question becomes more along the lines of, is readability mostly important for those deep into the code, or for the general populous.

As I personally contribute to a lot of different repositories, and am not very deeply into most projects I contribute to, I can tell you for a fact, having more explicit code is almost always easier to follow.

> Not true, in some cases maybe. But the context can be lost as well, as we have lots of modules that have the same name. For example, a recorder can be an entity platform implementation too. The context doesn't tell you the origin of recorder in the given example.

Hmm, so because it's not always true, we shouldn't do it at all? I think reducing the context would be already a big win.

> Same as above. It doesn't solve it. There are many tools that can help you a lot better with that instead of searching. That said, you can still search for the import in the current situation.
IDEs will pick up the references as well.

Sure; though having to rely on an IDE to work efficiently is also a personal preference :p

> This is a non-reason. IDEs have no issues with the existing method either, as a matter of fact, they will even update the imports for you nowadays (and even clean them up when you are no longer using an imported reference).

Ok cool, though personally I'm not a fan of putting all my eggs in the IDE basket :p

> PEP020 is the biggest &% Python ever produced

Hmm, to me however, it seems quite reasonable to be fair. Just because their implementation doesn't follow their own rules to the letter ...


Anyway, as I stated, style can become quite personal, and I sense very strongly that this is personal to you :) and that's okay. We may end up just having to agree to disagree, as I obviously think it IS a good idea ;)

But please document your preferred style and lets make the codebase follow it, having a linter would then also help. Right now, both variants and a  bit are being done right now, which is inconsistent. Though I suppose better naming on a few things would also help a lot.


> I don't like importing complete package, I prefer to know what is used from the package when looking at the imports.

I generally agree with that, but should we not do this always anyway? And if we do, should we not have better names maybe? e.g. always use `as <module>_function` (which then of course becomes your import documentation, which would in the end be almost the same as `<module>.function`. Some modules can be quite big (which might be considered bad) but having to keep a constant mind-map of the imports is of course annoying.

**Reply by Jc2k** [reply]
Architecture issues are generally where you discuss changes that you plan to make, rather than for requesting work to be done.

Are you planning to update our ruff/pylint/mypy/flake etc config based on the outcome of this conversation?

What other changes are you going to make?

Do you have any numbers on how many modules do it one way vs the other?

**Reply by oliv3r** [reply]
> Architecture issues are generally where you discuss changes that you plan to make, rather than for requesting work to be done.

I did the work, and was asked to create an architecture discussion about the topic, not sure where you are reading that I'm requesting work to be done?

> Are you planning to update our ruff/pylint/mypy/flake etc config based on the outcome of this conversation?

Lets first get to a conclusion of the discussion. Defining what is desired/preferred and take it from there?

> What other changes are you going to make?

Other changes compared to? I was just requesting work to be done (jk). I did it for the recorder, and I could do it for other components as well of course if it helps and improves things.

> Do you have any numbers on how many modules do it one way vs the other?

In the entire world? That's quite hard to say and also quite skewed. If (depending on what is considered better) most projects that would be scanned are sloppy quick hacks, this skews your numbers pretty badly right? So you'd only have to do it for high-quality python projects.
If you speak of only hass, it's a mixed back 'sometimes you see the one, sometimes the other'. I would say what I've seen so far, is that the number of imports that's importing specific methods/functions is somewhat less, but that's just based on me remembering. But then what does this data tell us, other then what is favored by the maintainers (which is okay of course). But rather then analyzing what is favored, it's better to discuss and nail it down, put it in words, so it's always clear for everyone else.

**Reply by Jc2k** [reply]
Sorry I didn't see the link to your closed PR, and it was probably the most important bit of that whole post as well 😭

I was more interested on stats within HA. Big changes normally need breaking up into many smaller PRs. that's how we deal with incremental code quality improvements. Smaller changes are easier to review. That's part of the planning we need to do before doing a big PR.

I personally think it's not that big of a deal as things are. If there's a case where it's genuinely confusing or bad practice the maintainers can and do ask for changes in review. We already have quite a few linters and they have lifted our code quality and maintainability a lot. I don't think changing the import style is going to be as impactful for code quality as black, mypy and ruff in particular have been. Or pylint or any of the linters we already use.

With that in mind I personally would only enforce rules around it if they could be enforced by one of our linters. Ideally ruff or one that can auto fix. Definitely one that can integrate with the on commit hook that we use. That's why I asked if you could provide a config for one that could. If you can't, I'm against changing anything. 

On a large code base where the PR to reviewer ratio is skewed like it is, I feel like it's almost busy work for the maintainers to manually enforce something like this. HA is a complex codebase and I'd rather have maintainers use their expertise to guide contributors towards better use of internal APIs that they are experts in. Unless it's critical to the maintainability of the code base, I personally think it's a waste of their precious time. And I really don't think this change alone is critical, either the original or the counter proposal. (This doesn't mean they can't comment on this topic if they feel it's problematic in a given PR).

I don't know to what extent you want to document this, but fwiw I'm also personally against a style guide that perfectly duplicates E.g our ruff/flake/whatever rules. The configuration is the documentation. I don't think the HA project should use resources that could be spent fixing bugs to keep a document up to date with the latest changes to the lint configuration. This might be a little newbie hostile, but ideally they can run ruff --fix so they can write in their style and get HA style before they commit. I have never seen a black style guide but it only took a few PRs before my internalised coding style was 95% black. Not that it matters, because it would fix my style regardless.

**Reply by oliv3r** [reply]
> Sorry I didn't see the link to your closed PR, and it was probably the most important bit of that whole post as well sob

it happens :D no hard feelings


> I was more interested on stats within HA. Big changes normally need breaking up into many smaller PRs. that's how we deal with incremental code quality improvements. Smaller changes are easier to review. That's part of the planning we need to do before doing a big PR.

Yeah, the PR kind of happened, each time adding things. I won't say it should be split up; we can easily split it; but in essence, it technically IS one change, over the entire codebase. Splitting it into multiple commits (one per module) is feasible. Doing multple MR's is a bit ... weird, but possible.

> 
> I personally think it's not that big of a deal as things are. If there's a case where it's genuinely confusing or bad practice the maintainers can and do ask for changes in review. We already have quite a few linters and they have lifted our code quality and maintainability a lot. I don't think changing the import style is going to be as impactful for code quality as black, mypy and ruff in particular have been. Or pylint or any of the linters we already use.

I think this is a very valid point. It needs to be possible to configure a linter to do so, if this isn't possible, it's a lot of manual work. However, realize, that even the current solution, IS manual work. Is every maintainer telling people how it should be done when they do it in a nother way? (the answer is probably no, because currently, it's a mixed bag).

> With that in mind I personally would only enforce rules around it if they could be enforced by one of our linters. Ideally ruff or one that can auto fix. Definitely one that can integrate with the on commit hook that we use. That's why I asked if you could provide a config for one that could. If you can't, I'm against changing anything.

I can agree with that sentiment, however we can still write down a preference, so that it can be referred to against?

> On a large code base where the PR to reviewer ratio is skewed like it is, I feel like it's almost busy work for the maintainers to manually enforce something like this. HA is a complex codebase and I'd rather have maintainers use their expertise to guide contributors towards better use of internal APIs that they are experts in. Unless it's critical to the maintainability of the code base, I personally think it's a waste of their precious time. And I really don't think this change alone is critical, either the original or the counter proposal. (This doesn't mean they can't comment on this topic if they feel it's problematic in a given PR).

Cannot argue against this valid point either :)

> I don't know to what extent you want to document this, but fwiw I'm also personally against a style guide that perfectly duplicates E.g our ruff/flake/whatever rules. The configuration is the documentation. I don't think the HA project should use resources that could be spent fixing bugs to keep a document up to date with the latest changes to the lint configuration. This might be a little newbie hostile, but ideally they can run ruff --fix so they can write in their style and get HA style before they commit. I have never seen a black style guide but it only took a few PRs before my internalised coding style was 95% black. Not that it matters, because it would fix my style regardless.

Right now, ruff 'ignores' import style, anything goes. But I'll see what the linters can do here, especially ruff.

**Comment by Jc2k** [comment]
> Doing multple MR's is a bit ... weird, but possible.

It's not weird here :-) HA is a large project, i think its one of the largest/busiest on GitHub? It has lots of contributors responsible for their own little bits. The experience is again and again (at least as i understand it) that the larger a PR is, the less effective the review is. It is simply easier to be confident that a small PR is correct. This is less of an issue with something like `black` or `ruff --fix` - but if a human has made a 5000 line diff there is a higher chance that there is a typo, and if its 50 copies of the same 100 line change, its very easy to miss when scrolling through a large diff. To mitigate these sorts of challenges changes are broken up. How they are broken up... theres no hard rules. Its pragmatic, and depends on how many files are touched.

PR reviewers are assigned based on components (`homeassistant/components/*/*.py`) touched. Looking at my dev env, it has 1174 "components" (most of these are integrations). I am personally responsible for 3 of those, and can approve pull requests for those. Ideally for me as a maintainer, I would get PR's for my pieces and be able to approve and merge them without having to wade through another 5k lines. But in that case in the worst case you might have to open 1174 pull requests. That is obviously not workable either. So the current approach is a trade off - to increase reviewability, without putting too much burden on contributors either.

> However, realize, that even the current solution, IS manual work. Is every maintainer telling people how it should be done when they do it in a nother way? (the answer is probably no, because currently, it's a mixed bag).

I think you hit the nail on the head with the last sentence. There is no "current solution" with regards this arch discussion. They are reviewing API use and they are reviewing implementation. I have 221 PR's against HA, and the review here is some of the most in depth, thorough and strict i've encountered. Reviewers here are doing that review and doing it well. They are patient and insightful, and I can say my own integrations are far far better for the care they have shown. So, if a maintainer feels that the import style is a massive issue in terms of readability or maintainability, they would have been saying. It seems they are generally not, hence the inconsistenty. I can certainly say for the PR's I have reviewed, beyond the imports being valid and isort passing, I haven't considered the imports. I have not ever found them problematic from a maintability point of view. So ultimately asking for this change to be followed on an ongoing basis by all maintainers for every integration is a lot of new work that wasn't being done before. And some of us will probably forget.

**Reply by oliv3r** [reply]
> It is simply easier to be confident that a small PR is correct. This is less of an issue with something like black or ruff --fix - but if a human has made a 5000 line diff there is a higher chance that there is a typo,

This is absolutely true, my statement was more about the 'change' being the same for 5000 lines. So then it becomes a bit weird, not impossible of course.

> I am personally responsible for 3 of those, and can approve pull requests for those

I think this is a key point for having 50 MR's over 50 commits in a single MR, I hadn't looked at it from that pov.! Thanks for sharing.

Using tooling however to do the same, is much more preferred for sure!

> And some of us will probably forget.

Also very much understood. But having down a 'guideline' at least means there's a preferred way. And once `ruff` can start to fix it, and check on it, the preferred way can be automatically followed, without much thought.


But these last posts have been very insightful, albeit 'beating around the bush' :) Let me look into what's possible with tooling first.
